adoc/admin-gpus.adoc

```diff
@@ -181,3 +181,143 @@ Check the kubelet log on the worker node that has a GPU. This may indicate error
 ----
 # journalctl -u kubelet
 ----
+
+== Monitoring
+
+If you have configured link:{docurl}/single-html/caasp-admin/#_monitoring[Monitoring] for your cluster, you may want to use {nvidia}'s Data Center GPU Manager (DCGM) to monitor your GPUs. Follow the steps below to configure the Prometheus exporter and Grafana dashboard for your {nvidia} GPUs.
+
+=== Configure a Pod Security Policy
+
+The DCGM requires use of the hostPath volume type to access the kubelet socket on the host worker node. Create an appropriate Pod Security Policy and RBAC configuration to allow this:
+
+
+[source,bash]
+----
+$ cat <<EOF | kubectl apply -f -
+---
+apiVersion: policy/v1beta1
+kind: PodSecurityPolicy
+metadata:
+  name: nvidia.dcgm
+spec:
+  privileged: false
+  seLinux:
+    rule: RunAsAny
+  supplementalGroups:
+    rule: RunAsAny
+  runAsUser:
+    rule: RunAsAny
+  fsGroup:
+    rule: RunAsAny
+  allowedHostPaths:
+    - pathPrefix: /var/lib/kubelet/pod-resources
+  volumes:
+  - hostPath
+  - configMap
+  - secret
+  - emptyDir
+  - downwardAPI
+  - projected
+  - persistentVolumeClaim
+  - nfs
+  - rbd
+  - cephFS
+  - glusterfs
+  - fc
+  - iscsi
+  - cinder
+  - gcePersistentDisk
+  - awsElasticBlockStore
+  - azureDisk
+  - azureFile
+  - vsphereVolume
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: nvidia:dcgm
+rules:
+- apiGroups:
+  - policy
+  resources:
+  - podsecuritypolicies
+  verbs:
+  - use
+  resourceNames:
+  - nvidia.dcgm
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRoleBinding
+metadata:
+  name: nvidia:dcgm
+roleRef:
+  kind: ClusterRole
+  name: nvidia:dcgm
+  apiGroup: rbac.authorization.k8s.io
+subjects:
+- kind: Group
+  name: system:serviceaccounts:dcgm
+---
+----
+
+=== Create the DCGM Exporter
+
+The DCGM exporter monitors GPUs on each worker node and exposes metrics that can be queried.
+
+[source,bash]
+----
+$ kubectl create namespace dcgm
+$ kubectl create --namespace dcgm -f https://raw.githubusercontent.com/NVIDIA/gpu-monitoring-tools/master/dcgm-exporter.yaml
+----
+
+Check that the metrics are being collected:
+
+[source,bash]
+----
+$ NAME=$(kubectl get pods --namespace dcgm -l "app.kubernetes.io/name=dcgm-exporter" -o "jsonpath={ .items[0].metadata.name}")
+$ kubectl port-forward $NAME 8080:9400 &
```

Comment:
Is this intentionally backgrounded?  If so, should the job be killed later on, or just left running in the background?  Does it exit after the connection is finished?  I've never used `port-forward`. :)

Reply:
I do like breaking this up into multiple lines using the shell variable.  :+1:

Reply:
This is copypasted from the upstream documentation, probably to avoid needing to change your terminal after entering the command. It seems to die on its own after a while.

Reply:
I'm back, and better informed! :D
According to https://kubernetes.io/docs/reference/generated/kubectl/kubectl-commands#port-forward, the forwarding remains until the connected-to pod terminates.  So this would otherwise hang around until the shell session ends.

As such, it might not be bad to provide a command to stop the backgrounded port forwarding.  Either `kill %kubectl` or `kill %+` would be good options.  Those kill a backgrounded kubectl (if there's only one) or the most recent backgrounded job, respectively - I think I prefer the `%+` here.  Otherwise there's maybe a potential security hole leaving a port open, and there's the annoying "there are stopped jobs" when you try to exit the shell.

I'm not really tied to that needing to be there, though, so it's up to you if adding the cleanup is worth the extra clutter.

Reply:
I think this is getting too into the weeds, this is just supposed to be a quick example to check the functionality of the metrics server, not a lesson in bash job control. Moreover I think it's dangerous to instruct users to use `kill` when we don't know for sure what other processes they may have stopped in between running these examples. I can remove the `&` if that would put you more at ease?

Reply:
If you remove the `&`, the forward stays in the foreground, which requires opening another terminal. That seems worse.

It seems pretty unlikely that the user would start the forwarding, do the curl, and then take a moment to restart the enterprise payroll database in the same shell session before finishing the example. But maybe I'm overestimating the target audience. ;)

Like I said, I'm not married to the cleanup command, so if that's concerning, I approve of this as-is as well. 👍

Reply:
> If you remove the &, the forward stays in the foreground, which requires opening another terminal.

Yes, but then it's deliberate, and then they either ^C the port-forward or close that terminal window when they are done, and your concern seems to be that we're leaving garbage processes running without the user realizing it.

> That seems worse.

Worse from a security perspective, or a user-friendliness perspective?

> It seems pretty unlikely that the user would start the forwarding, do the curl, and then take a moment to restart the enterprise payroll database in the same shell session before finishing the example.

It doesn't have to be that drastic, maybe they start the forwarding, open vim to look up some notes, then ^Z their vim session, then `kill %+` would kill that and leave an orphaned .swp file as well as leave them in the same position of having the port remain open.

Reply:
A separate terminal would work fine too. That's common in tutorials, so people probably wouldn't think twice about it. There's something to be said for simplicity.

Should we want a bit more sophistication, since we _are_ the experts ;) - the `%kubectl` format I mentioned addresses the "wrong process" issue. It can contain the whole quoted command line if need be for further disambiguation, like `%"kubectl port-forward"` - which is also pretty descriptive of what's happening.

Either way, I still might argue that there could be a valuable lesson hidden here for the one person who reads our docs, backgrounds commands a lot, and blindly copy-pastes commands. 😂

Reply:
Verbose rambling and apparent bias aside, I really do think both options are equivalent and both preferable to just leaving it.

Reply:
Like I said, I used the `&` in keeping with the [upstream documentation](https://github.com/nvidia/gpu-monitoring-tools#quickstart-on-kubernetes) for this tool, not because I don't know how linux job control works. In looking for other examples online, I don't see anyone using `&`, they instead just opt to ignore the fact that the port-forward command takes over your terminal. I've changed it to remove the `&` and explicitly instruct the user to open a separate terminal. I think this is simplest and doesn't leave anyone guessing about the meaning of obscure kill syntax. I really hope we can leave it at this.